File: sole-prop-extra-numbers-unregistered/python/twilio_sole_prop_pool_audit.py
"""Find Sole Proprietor Messaging Services holding more than one sender.

Read only. GET requests and nothing else: give this an API Key with read access
rather than the account auth token. Detaching the wrong number here would take
the one registered sender out of the pool, so the removals are printed and a
person decides which number stays.
"""
import argparse
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("twilio_sole_prop_pool_audit")

MSG = "https://messaging.twilio.com/v1"

SOLE = "SOLE_PROPRIETOR"

# A Sole Proprietor brand permits one campaign, and that campaign one 10DLC
# number. The limit lives on the brand, not on the Messaging Service.
SOLE_PROP_SENDER_LIMIT = 1


def verdict(brand, pool_size, campaign_status=None, limit=SOLE_PROP_SENDER_LIMIT):
    """Classify one Messaging Service against its brand's sender limit.

    `brand` is the brand registration dict or None, `pool_size` the number of
    phone numbers in the service's pool or None. Nothing is fetched here, so
    every state below is testable offline. Returns (state, detail).
    """
    if brand is None:
        return ("brand-unread",
                "the campaign names a brand_registration_sid that could not be "
                "read, so the one sender limit cannot be applied to this pool.")

    brand_type = str(brand.get("brand_type") or "").upper()
    if brand_type != SOLE:
        return ("not-sole-prop",
                "brand_type is %s: the pool size is not capped by the brand."
                % (brand_type or "unset"))

    if pool_size is None:
        return ("pool-unread",
                "sole proprietor brand and the sender pool could not be read.")

    status = str(campaign_status or "").upper()

    if pool_size == 0:
        return ("empty-pool",
                "sole proprietor brand with nothing in the sender pool. Every "
                "US send fails consistently rather than intermittently, and "
                "the repair is to add the one number rather than remove any.")

    if pool_size > limit:
        extras = pool_size - limit
        return ("overfilled",
                "%d numbers in the pool on a sole proprietor brand, which "
                "permits %d. %d of them will sit at A2P status UNREGISTERED "
                "permanently, and the service picks a sender per message, so "
                "30034 arrives at random rather than for one from."
                % (pool_size, limit, extras))

    if status and status != "VERIFIED":
        return ("single-not-verified",
                "one number, which is the limit, but campaign_status is %s so "
                "it is not registered yet. This is the review clock, not the "
                "sender limit." % status)

    return ("registered",
            "one number in the pool, which is what a sole proprietor brand "
            "supports.")


def get(session, url, **params):
    r = session.get(url, params=params, timeout=30)
    if r.status_code in (401, 403):
        raise SystemExit("%d from Twilio: check TWILIO_ACCOUNT_SID and that the "
                         "API key belongs to that account with read access"
                         % r.status_code)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def list_v1(session, url, key, limit=1000):
    """Page a messaging.twilio.com list. meta.next_page_url is absolute."""
    out = []
    while url and len(out) < limit:
        page = get(session, url, PageSize=50)
        if page is None:
            break
        out.extend(page.get(key, []))
        url = (page.get("meta") or {}).get("next_page_url")
    return out[:limit]


def read_brand(session, cache, brand_sid):
    """Read a brand once per run. Several services can share one brand."""
    if not brand_sid:
        return None
    if brand_sid not in cache:
        cache[brand_sid] = get(session, "%s/a2p/BrandRegistrations/%s"
                               % (MSG, brand_sid))
    return cache[brand_sid]


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--max-services", type=int, default=200)
    args = ap.parse_args()

    account = os.environ.get("TWILIO_ACCOUNT_SID")
    key = os.environ.get("TWILIO_API_KEY")
    secret = os.environ.get("TWILIO_API_SECRET")
    if not (account and key and secret):
        log.error("set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET "
                  "(an API Key with read access, not the auth token)")
        return 2

    session = requests.Session()
    session.auth = (key, secret)
    brands = {}

    services = list_v1(session, MSG + "/Services", "services", args.max_services)
    if not services:
        log.info("no Messaging Services on this account")
        return 0

    sole = 0
    bad = 0
    for svc in services:
        campaigns = list_v1(session, "%s/Services/%s/Compliance/Usa2p"
                            % (MSG, svc["sid"]), "compliance")
        campaign = campaigns[0] if campaigns else None
        if campaign is None:
            continue
        brand = read_brand(session, brands, campaign.get("brand_registration_sid"))
        numbers = list_v1(session, "%s/Services/%s/PhoneNumbers" % (MSG, svc["sid"]),
                          "phone_numbers")
        state, detail = verdict(brand, len(numbers),
                                campaign.get("campaign_status"))
        if state == "not-sole-prop":
            continue
        sole += 1
        name = svc.get("friendly_name") or svc["sid"]
        line = "%-20s %s  %s" % (state, name, detail)
        if state == "registered":
            log.info(line)
            continue
        bad += 1
        log.warning(line)
        if state == "overfilled":
            for num in numbers:
                log.warning("    in pool: %s  %s", num.get("phone_number", "?"),
                            num.get("sid", ""))
            log.warning("  repair: detach every number above except the one that "
                        "is actually registered, at %s/Services/%s/PhoneNumbers/"
                        "{PhoneNumberSid}. Confirm which one is registered "
                        "first: removing the wrong two turns an intermittent "
                        "failure into a total one", MSG, svc["sid"])
            log.warning("  repair: if this account genuinely needs more senders, "
                        "register a Standard or Low-Volume Standard brand. Sole "
                        "Proprietor cannot be widened")
        elif state == "empty-pool":
            log.warning("  repair: attach the intended sender to %s, then wait "
                        "for its A2P registration to complete", svc["sid"])

    log.info("%d service(s), %d on a sole proprietor brand, %d overfilled",
             len(services), sole, bad)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())

File: sole-prop-extra-numbers-unregistered/node/twilio-sole-prop-pool-audit.mjs
/**
 * Find Sole Proprietor Messaging Services holding more than one sender.
 *
 * Read only. GET requests and nothing else: give this an API Key with read
 * access rather than the account auth token. The removals are printed, because
 * detaching the wrong number takes the one registered sender out of the pool.
 */
const MSG = 'https://messaging.twilio.com/v1';

const SOLE = 'SOLE_PROPRIETOR';

// A Sole Proprietor brand permits one campaign, and that campaign one 10DLC
// number. The limit lives on the brand, not on the Messaging Service.
export const SOLE_PROP_SENDER_LIMIT = 1;

/**
 * Classify one Messaging Service against its brand's sender limit. Nothing is
 * fetched here. Pure. Returns [state, detail].
 */
export function verdict(brand, poolSize, campaignStatus = null,
                        limit = SOLE_PROP_SENDER_LIMIT) {
  if (brand === null || brand === undefined) {
    return ['brand-unread',
      'the campaign names a brand_registration_sid that could not be read, so ' +
      'the one sender limit cannot be applied to this pool.'];
  }

  const brandType = String(brand.brand_type ?? '').toUpperCase();
  if (brandType !== SOLE) {
    return ['not-sole-prop',
      `brand_type is ${brandType || 'unset'}: the pool size is not capped by the brand.`];
  }

  if (poolSize === null || poolSize === undefined) {
    return ['pool-unread',
      'sole proprietor brand and the sender pool could not be read.'];
  }

  const status = String(campaignStatus ?? '').toUpperCase();

  if (poolSize === 0) {
    return ['empty-pool',
      'sole proprietor brand with nothing in the sender pool. Every US send ' +
      'fails consistently rather than intermittently, and the repair is to add ' +
      'the one number rather than remove any.'];
  }

  if (poolSize > limit) {
    const extras = poolSize - limit;
    return ['overfilled',
      `${poolSize} numbers in the pool on a sole proprietor brand, which ` +
      `permits ${limit}. ${extras} of them will sit at A2P status UNREGISTERED ` +
      'permanently, and the service picks a sender per message, so 30034 ' +
      'arrives at random rather than for one from.'];
  }

  if (status && status !== 'VERIFIED') {
    return ['single-not-verified',
      `one number, which is the limit, but campaign_status is ${status} so it ` +
      'is not registered yet. This is the review clock, not the sender limit.'];
  }

  return ['registered',
    'one number in the pool, which is what a sole proprietor brand supports.'];
}

function authHeader(key, secret) {
  return `Basic ${Buffer.from(`${key}:${secret}`).toString('base64')}`;
}

async function get(auth, url, params = {}) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  const res = await fetch(u, { headers: { Authorization: auth } });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from Twilio: check TWILIO_ACCOUNT_SID and ` +
                    'that the API key belongs to that account with read access');
  }
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`${res.status} from ${u.pathname}`);
  return res.json();
}

export async function listV1(auth, url, key, limit = 1000) {
  const out = [];
  let next = url;
  while (next && out.length < limit) {
    const page = await get(auth, next, { PageSize: 50 });
    if (page === null) break;
    out.push(...(page[key] ?? []));
    next = page.meta?.next_page_url ?? null;
  }
  return out.slice(0, limit);
}

async function readBrand(auth, cache, brandSid) {
  if (!brandSid) return null;
  if (!(brandSid in cache)) {
    cache[brandSid] = await get(auth, `${MSG}/a2p/BrandRegistrations/${brandSid}`);
  }
  return cache[brandSid];
}

async function main() {
  const account = (process.env.TWILIO_ACCOUNT_SID || "dummy-twilio-account-sid");
  const key = (process.env.TWILIO_API_KEY || "dummy-twilio-api-key");
  const secret = (process.env.TWILIO_API_SECRET || "dummy-twilio-api-secret");
  if (!account || !key || !secret) {
    console.error('set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET ' +
                  '(an API Key with read access, not the auth token)');
    process.exitCode = 2;
    return;
  }
  const auth = authHeader(key, secret);
  const brands = {};

  const services = await listV1(auth, `${MSG}/Services`, 'services');
  if (services.length === 0) {
    console.log('no Messaging Services on this account');
    return;
  }

  let sole = 0;
  let bad = 0;
  for (const svc of services) {
    const campaigns = await listV1(auth, `${MSG}/Services/${svc.sid}/Compliance/Usa2p`,
                                   'compliance');
    const campaign = campaigns[0] ?? null;
    if (campaign === null) continue;
    const brand = await readBrand(auth, brands, campaign.brand_registration_sid);
    const numbers = await listV1(auth, `${MSG}/Services/${svc.sid}/PhoneNumbers`,
                                 'phone_numbers');
    const [state, detail] = verdict(brand, numbers.length, campaign.campaign_status);
    if (state === 'not-sole-prop') continue;
    sole += 1;
    const name = svc.friendly_name ?? svc.sid;
    const line = `${state.padEnd(20)} ${name}  ${detail}`;
    if (state === 'registered') { console.log(line); continue; }
    bad += 1;
    console.warn(line);
    if (state === 'overfilled') {
      for (const num of numbers) {
        console.warn(`    in pool: ${num.phone_number ?? '?'}  ${num.sid ?? ''}`);
      }
      console.warn('  repair: detach every number above except the one that is ' +
                   `actually registered, at ${MSG}/Services/${svc.sid}/` +
                   'PhoneNumbers/{PhoneNumberSid}. Confirm which one is ' +
                   'registered first: removing the wrong two turns an ' +
                   'intermittent failure into a total one');
      console.warn('  repair: if this account genuinely needs more senders, ' +
                   'register a Standard or Low-Volume Standard brand. Sole ' +
                   'Proprietor cannot be widened');
    } else if (state === 'empty-pool') {
      console.warn(`  repair: attach the intended sender to ${svc.sid}, then ` +
                   'wait for its A2P registration to complete');
    }
  }

  console.log(`${services.length} service(s), ${sole} on a sole proprietor brand, ` +
              `${bad} overfilled`);
  process.exitCode = bad ? 1 : 0;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing credentials and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
